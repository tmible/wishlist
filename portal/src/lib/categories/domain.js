/** @module Домен категорий */

/**
 * Категория элементов списка желаний пользователя
 * @typedef {object} Category
 * @property {number} id Идентификатор категории
 * @property {string} name Название категории
 */

/**
 * Проверка наличия различий в двух экземплярах категории.
 * Оба объекта должны быть экземплярами одной и той же категории
 * @function isCategoryEdited
 * @param {Category} category Оригинальный экземпляр категории
 * @param {Category} editedCategory Изменённый экземпляр категории
 * @returns {boolean} Признак наличия различий
 */
export const isCategoryEdited = (
  category,
  editedCategory,
) => category.id === editedCategory.id && category.name !== editedCategory.name;
