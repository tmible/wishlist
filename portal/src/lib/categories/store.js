import { initCategories } from '$lib/categories/use-cases/init-categories.js';

/** @typedef {import('./domain.js').Category} Category */

/** @module Адаптер хранилища категорий */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {Category[]}
 */
let value = [];

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * категорий
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => Category[]} get Получение значения хранилища
 * @property {(newValue: Category[]) => void} set Установка значения хранилища
 * @property {(category: Category) => void} add Добавление категории в хранилище
 * @property {(category: Category) => void} update Обновление категории в хранилище
 * @property {(category: Category) => void} delete Удаление категории из хранилища
 */
export const categories = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);

    if (subscriptions.size === 1) {
      initCategories();
    }

    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  add: (category) => {
    value.push(category);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  update: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value[categoryIndex] = category;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },

  delete: (category) => {
    const categoryIndex = value.findIndex(({ id }) => id === category.id);
    if (categoryIndex === -1) {
      return;
    }
    value.splice(categoryIndex, 1);
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
