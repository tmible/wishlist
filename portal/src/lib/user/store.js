import { DEFAULT_USER } from './default-user.const.js';

/** @typedef {import('./domain.js').User} User */

/** @module Адаптер хранилища пользователя */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {User}
 */
let value = DEFAULT_USER;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * пользователя
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => User} get Получение значения хранилища
 * @property {(newValue: User) => void} set Установка значения хранилища
 */
export const user = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },
  get: () => value,
  set: (newValue) => {
    value = newValue;
    for (const subscriber of subscriptions.values()) {
      subscriber(value);
    }
  },
};
