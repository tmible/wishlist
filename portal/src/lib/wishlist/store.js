/** @typedef {import('./domain.js').OwnWishlistItem} OwnWishlistItem */
/** @typedef {import('./use-cases/reorder-list.js').ReorderPatch} ReorderPatch */

/** @module Адаптер хранилища списка желаний */

/** @typedef {(value: any) => void} Subscriber */
/** @typedef {() => void} Unsubscriber */

/**
 * Значение хранилица
 * @type {OwnWishlistItem[] | null}
 */
let value = null;

/**
 * Подписчики хранилища
 * @type {Map<Subscriber, Subscriber>}
 */
const subscriptions = new Map();

/**
 * Уведомление подписчиков об изменении значения хранилища
 * @function notifySubscribers
 * @returns {void}
 */
const notifySubscribers = () => {
  for (const subscriber of subscriptions.values()) {
    subscriber(value);
  }
};

/**
 * [Svelte-совместимое хранилище]{@link https://svelte.dev/docs/svelte/stores#Store-contract}
 * списка желаний
 * @type {object} Store
 * @property {(subscriber: Subscriber) => Unsubscriber} subscribe Подписка на значение хранилища
 * @property {() => OwnWishlistItem[]} get Получение значения хранилища
 * @property {(newValue: OwnWishlistItem[]) => void} set Установка значения хранилища
 * @property {(item: OwnWishlistItem) => void} add Добавление элемента списка в хранилище
 * @property {(item: OwnWishlistItem) => void} update Обновление элемента списка в хранилище
 * @property {(itemIds: OwnWishlistItem['id'][]) => void} delete
 *   Удаление элементов списка из хранилища
 * @property {(patch: ReorderPatch) => void} reorder Переупорядочивание списка в хранилище
 */
export const wishlist = {
  subscribe: (subscriber) => {
    subscriber(value);
    subscriptions.set(subscriber, subscriber);
    return () => subscriptions.delete(subscriber);
  },

  get: () => value,

  set: (newValue) => {
    value = newValue;
    notifySubscribers();
  },

  add: (item) => {
    if (!value) {
      value = [];
    }
    value.push(item);
    notifySubscribers();
  },

  update: (item) => {
    const itemIndex = value?.findIndex(({ id }) => id === item.id);
    if (itemIndex === -1) {
      return;
    }
    value[itemIndex] = item;
    notifySubscribers();
  },

  delete: (itemIds) => {
    let isAnyDeleted = false;

    for (const id of itemIds) {
      const itemIndex = value?.findIndex((item) => id === item.id);
      if (itemIndex === -1) {
        continue;
      }
      value.splice(itemIndex, 1);
      isAnyDeleted = true;
    }

    if (!isAnyDeleted) {
      return;
    }

    notifySubscribers();
  },

  reorder: (patch) => {
    if (!value) {
      return;
    }
    for (const item of value) {
      const reorder = patch.find(({ id }) => id === item.id);
      item.order = reorder?.order ?? item.order;
    }
    value.sort((a, b) => a.order - b.order);
    notifySubscribers();
  },
};
