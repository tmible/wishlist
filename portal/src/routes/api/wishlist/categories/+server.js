import { json } from '@sveltejs/kit';
import { inject } from '@tmible/wishlist-common/dependency-injector';
import { InjectionToken } from '$lib/architecture/injection-token';

/**
 * Получение категорий пользователя
 * @type {import('./$types').RequestHandler}
 */
export const GET = ({ locals }) => {
  const { userid } = locals;
  return json(inject(InjectionToken.GetUserCategoriesStatement).all(userid));
};

/**
 * Добавление категории
 * @type {import('./$types').RequestHandler}
 */
export const POST = async ({ request, locals }) => {
  const { userid } = locals;
  const name = await request.text();
  inject(InjectionToken.AddCategoryStatement).run(userid, name);
  return new Response(null, { status: 200 });
};
