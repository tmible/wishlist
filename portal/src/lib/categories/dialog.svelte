<!-- @component Диалог с категориями -->
<script>
  import ScrollArea from '@tmible/wishlist-ui/scroll-area';
  import { Dialog } from 'bits-ui';
  import Pencil from 'lucide-svelte/icons/pencil';
  import Save from 'lucide-svelte/icons/save';
  import Trash2 from 'lucide-svelte/icons/trash-2';
  import Undo2 from 'lucide-svelte/icons/undo-2';
  import X from 'lucide-svelte/icons/x';
  import { categories } from './store.js';
  import { createCategory } from './use-cases/create-category.js';
  import { deleteCategory as deleteCategoryUseCase } from './use-cases/delete-category.js';
  import { editCategory as editCategoryUseCase } from './use-cases/edit-category.js';

  /** @typedef {import('./domain.js').Category} Category */

  /**
   * @typedef {object} Props
   * @property {boolean} [open] Признак открытости диалога
   */

  /** @type {Props} */
  let { open = $bindable(false) } = $props();

  /**
   * Ссылка на элемент с содержимым диалога
   * @type {HTMLElement}
   */
  let content = $state(null);

  /**
   * Значение инпута ввода названия новой категории
   * @type {string | null}
   */
  let newCategoryName = $state(null);

  /**
   * Редактируемая категория
   * @type {Category | null}
   */
  let editingCategory = $state(null);

  /**
   * Инпут ввода названия редактируемой категории. Используется для установки фокуса
   * @type {HTMLInputElement}
   */
  let editingCategoryInput = $state(null);

  /**
   * Обработчк изменения открытости диалога. Сбрасывает состояние при закрытии диалога
   * @type {(open: boolean) => void}
   */
  const onOpenChange = (isOpen) => {
    if (isOpen) {
      return;
    }
    newCategoryName = null;
    editingCategory = null;
    editingCategoryInput = null;
  };

  /**
   * Добавление новой категории
   * @function addCategory
   * @returns {Promise<void>}
   * @async
   */
  const addCategory = async () => {
    const tmp = newCategoryName;
    newCategoryName = null;
    await createCategory({ name: tmp });
  };

  /**
   * Отмена редактирования категории
   * @function cancelEditingCategory
   * @returns {void}
   */
  const cancelEditingCategory = () => {
    editingCategory = null;
  };

  /**
   * Сохранение отредактированной категории
   * @function submitEditedCategory
   * @param {Category} category Редактируемая категория в начальном варианте
   * @returns {Promise<void>}
   * @async
   */
  const submitEditedCategory = async (category) => {
    const tmp = editingCategory;
    editingCategory = null;
    await editCategoryUseCase(category, tmp);
  };

  /**
   * Начало редактирования категории
   * @param {Category} category Редактируемвая категория
   * @returns {void}
   */
  const editCategory = (category) => {
    editingCategory = { ...category };
    setTimeout(() => editingCategoryInput.focus());
  };

  /**
   * Удаление категории
   * @param {Category} category Удаляемая категория
   * @returns {Promise<void>}
   * @async
   */
  const deleteCategory = async (category) => {
    await deleteCategoryUseCase(category);
  };
</script>

<Dialog.Root {onOpenChange} bind:open={open}>
  <Dialog.Trigger class="hidden" />
  <Dialog.Portal to="#modal-portal">
    <Dialog.Overlay class="modal-backdrop" />
    <Dialog.Content
      class="modal-box overflow-y-visible p-0"
      onOpenAutoFocus={() => setTimeout(() => content.focus())}
      bind:ref={content}
    >
      <ScrollArea class="max-h-[inherit]" viewportClasses="max-h-[inherit] p-6">
        <div class="flex items-start prose mb-3">
          <Dialog.Title class="mt-0 mr-auto">Категории</Dialog.Title>
          <Dialog.Close class="cursor-pointer">
            <X />
          </Dialog.Close>
        </div>
        <div class="flex items-center gap-4 mb-3">
          <input
            class="input w-full bg-base-200"
            type="text"
            placeholder="Новая категория"
            bind:value={newCategoryName}
          >
          <button
            class="btn btn-primary"
            class:btn-disabled={!newCategoryName}
            disabled={!newCategoryName}
            onclick={addCategory}
          >
            Добавить
          </button>
        </div>
        <div class="flex flex-col">
          {#each $categories as category (category.id)}
            <div
              class="flex items-center justify-between rounded-lg"
              class:hover:bg-base-200={editingCategory?.id !== category.id}
              class:pl-4={editingCategory?.id !== category.id}
            >
              {#if editingCategory?.id === category.id}
                <input
                  bind:this={editingCategoryInput}
                  class="input grow bg-base-200"
                  type="text"
                  placeholder={category.name}
                  bind:value={editingCategory.name}
                >
                <div>
                  <button
                    class="btn btn-ghost"
                    data-testid="cancel-button"
                    onclick={cancelEditingCategory}
                  >
                    <Undo2 />
                  </button>
                  <button
                    class="btn btn-ghost"
                    data-testid="submit-button"
                    onclick={() => submitEditedCategory(category)}
                  >
                    <Save />
                  </button>
                </div>
              {:else}
                {category.name}
                <div>
                  <button
                    class="btn btn-ghost"
                    data-testid="edit-button"
                    onclick={() => editCategory(category)}
                  >
                    <Pencil />
                  </button>
                  <button
                    class="btn btn-ghost"
                    data-testid="delete-button"
                    onclick={() => deleteCategory(category)}
                  >
                    <Trash2 />
                  </button>
                </div>
              {/if}
            </div>
          {/each}
        </div>
      </ScrollArea>
    </Dialog.Content>
  </Dialog.Portal>
</Dialog.Root>
