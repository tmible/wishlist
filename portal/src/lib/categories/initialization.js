import { deprive, provide } from '@tmible/wishlist-common/dependency-injector';
import { NetworkService, Store } from './injection-tokens.js';
import * as networkService from './network.service.js';
import { categories } from './store.js';

/**
 * Регистрация зависисмостей и подписка на события для работы с категориями
 * @function initCategoriesFeature
 * @returns {() => void} Функция освбождения зависисмостей и отписки от событий
 */
export const initCategoriesFeature = () => {
  provide(Store, categories);
  provide(NetworkService, networkService);

  return () => {
    deprive(Store);
    deprive(NetworkService);
  };
};
