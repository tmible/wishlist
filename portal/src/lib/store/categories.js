import { writable } from 'svelte/store';

/**
 * @template T
 * @typedef {import('svelte/store').Writable<T>} Writable
 */
/**
 * Категория элементов списка желаний пользователя
 * @typedef {object} Category
 * @property {number} id Идентификатор категории
 * @property {string} name Название категории
 */

/**
 * Svelte хранилище категорий элементов списка желаний пользователя
 * @type {Writable<Category[] | null>}
 */
export const categories = writable(null, (set, update) => {
  let timeout;

  update((value) => {
    if (value === null) {
      timeout = setTimeout(async () => set(await fetch(
        '/api/wishlist/categories',
      ).then(
        (response) => response.json(),
      )));
    }
    return value;
  });

  return () => timeout && clearTimeout(timeout);
});
