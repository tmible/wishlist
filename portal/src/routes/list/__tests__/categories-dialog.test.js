// @vitest-environment jsdom
import { act, cleanup, fireEvent, render, screen, waitFor } from '@testing-library/svelte';
import { userEvent } from '@testing-library/user-event';
import { writable } from 'svelte/store';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';
import { categories } from '$lib/store/categories.js';
import CategoriesDialog from '../categories-dialog.svelte';

vi.stubGlobal(
  'fetch',
  vi.fn().mockResolvedValue({ json: vi.fn(() => [{ id: 3, name: 'category 3' }]) }),
);
vi.mock(
  '$lib/store/categories.js',
  () => ({ categories: writable([{ id: 1, name: 'category 1' }, { id: 2, name: 'category 2' }]) }),
);

describe('categories dialog', () => {
  let user;
  let modalPortal;

  beforeEach(() => {
    modalPortal = document.createElement('div');
    modalPortal.id = 'modal-portal';
    document.body.append(modalPortal);
    user = userEvent.setup();
  });

  afterEach(() => {
    modalPortal.remove();
    vi.clearAllMocks();
    cleanup();
  });

  it('should be rendered closed', () => {
    expect(render(CategoriesDialog).baseElement).toMatchSnapshot();
  });

  it('should be rendered opened', () => {
    const { baseElement } = render(CategoriesDialog, { open: true });
    expect(baseElement).toMatchSnapshot();
  });

  describe('opened', () => {
    beforeEach(() => {
      render(CategoriesDialog, { open: true });
    });

    it('should set button disabled if input is empty', () => {
      const button = screen.getByText(/Добавить/, { selector: 'button' });
      expect(button.getAttribute('disabled')).toBe('');
    });

    describe('add new category', () => {
      beforeEach(async () => {
        vi.spyOn(vi.mocked(categories), 'set').mockImplementation(vi.fn());
        const input = screen.getByPlaceholderText(/Новая категория/);
        await act(() => fireEvent.input(input, { target: { value: 'new category' } }));
        const button = screen.getByText(/Добавить/, { selector: 'button' });
        await user.click(button);
      });

      it('should post new category', () => {
        expect(
          vi.mocked(fetch),
        ).toHaveBeenCalledWith(
          '/api/wishlist/categories',
          { method: 'POST', body: 'new category' },
        );
      });

      it('should refresh categories', () => {
        expect(vi.mocked(fetch)).toHaveBeenCalledWith('/api/wishlist/categories');
      });

      it('should set refreshed categories into store', async () => {
        await waitFor(
          () => expect(
            vi.mocked(categories.set),
          ).toHaveBeenCalledWith(
            [{ id: 3, name: 'category 3' }],
          ),
        );
      });
    });

    describe('edit category', () => {
      beforeEach(async () => {
        await user.click(screen.getAllByTestId(/edit-button/, { selector: 'button' })[0]);
      });

      it('should display input', () => {
        expect(screen.getByPlaceholderText('category 1')).toBeDefined();
      });

      it('should focus input', () => {
        expect(screen.getByPlaceholderText('category 1')).toBe(document.activeElement);
      });

      it('should hide edit button', () => {
        expect(screen.getAllByTestId(/edit-button/, { selector: 'button' }).length).toBe(1);
      });

      it('should hide delete button', () => {
        expect(screen.getAllByTestId(/delete-button/, { selector: 'button' }).length).toBe(1);
      });

      it('should display cancel button', () => {
        expect(screen.getAllByTestId(/cancel-button/, { selector: 'button' }).length).toBe(1);
      });

      it('should display submit button', () => {
        expect(screen.getAllByTestId(/submit-button/, { selector: 'button' }).length).toBe(1);
      });

      describe('on cancel', () => {
        beforeEach(async () => {
          await user.click(screen.getByTestId(/cancel-button/, { selector: 'button' }));
        });

        it('should hide input', () => {
          expect(screen.queryByPlaceholderText('category 1')).toBeNull();
        });

        it('should display edit button', () => {
          expect(screen.queryAllByTestId(/edit-button/, { selector: 'button' }).length).toBe(2);
        });

        it('should display delete button', () => {
          expect(screen.queryAllByTestId(/delete-button/, { selector: 'button' }).length).toBe(2);
        });

        it('should hide cancel button', () => {
          expect(screen.queryAllByTestId(/cancel-button/, { selector: 'button' }).length).toBe(0);
        });

        it('should hide submit button', () => {
          expect(screen.queryAllByTestId(/submit-button/, { selector: 'button' }).length).toBe(0);
        });
      });

      describe('on submit', () => {
        describe('if unchanged', () => {
          beforeEach(async () => {
            await user.click(screen.getByTestId(/submit-button/, { selector: 'button' }));
          });

          it('should hide input', () => {
            expect(screen.queryByPlaceholderText('category 1')).toBeNull();
          });

          it('should display edit button', () => {
            expect(screen.queryAllByTestId(/edit-button/, { selector: 'button' }).length).toBe(2);
          });

          it('should display delete button', () => {
            expect(screen.queryAllByTestId(/delete-button/, { selector: 'button' }).length).toBe(2);
          });

          it('should hide cancel button', () => {
            expect(screen.queryAllByTestId(/cancel-button/, { selector: 'button' }).length).toBe(0);
          });

          it('should hide submit button', () => {
            expect(screen.queryAllByTestId(/submit-button/, { selector: 'button' }).length).toBe(0);
          });
        });

        describe('if changed', () => {
          beforeEach(async () => {
            vi.spyOn(vi.mocked(categories), 'set').mockImplementation(vi.fn());
            await act(() => fireEvent.input(
              screen.getByDisplayValue('category 1'),
              { target: { value: 'category 1 new' } },
            ));
            await user.click(screen.getByTestId(/submit-button/, { selector: 'button' }));
          });

          it('should hide input', () => {
            expect(screen.queryByPlaceholderText('category 1')).toBeNull();
          });

          it('should display edit button', () => {
            expect(screen.queryAllByTestId(/edit-button/, { selector: 'button' }).length).toBe(2);
          });

          it('should display delete button', () => {
            expect(screen.queryAllByTestId(/delete-button/, { selector: 'button' }).length).toBe(2);
          });

          it('should hide cancel button', () => {
            expect(screen.queryAllByTestId(/cancel-button/, { selector: 'button' }).length).toBe(0);
          });

          it('should hide submit button', () => {
            expect(screen.queryAllByTestId(/submit-button/, { selector: 'button' }).length).toBe(0);
          });

          it('should put category', () => {
            expect(
              vi.mocked(fetch),
            ).toHaveBeenCalledWith(
              '/api/wishlist/categories/1',
              { method: 'PUT', body: 'category 1 new' },
            );
          });

          it('should refresh categories', () => {
            expect(vi.mocked(fetch)).toHaveBeenCalledWith('/api/wishlist/categories');
          });

          it('should set refreshed categories into store', async () => {
            await waitFor(
              () => expect(
                vi.mocked(categories.set),
              ).toHaveBeenCalledWith(
                [{ id: 3, name: 'category 3' }],
              ),
            );
          });
        });
      });
    });

    describe('delete category', () => {
      beforeEach(async () => {
        vi.spyOn(vi.mocked(categories), 'set').mockImplementation(vi.fn());
        await user.click(screen.getAllByTestId(/delete-button/, { selector: 'button' })[0]);
      });

      it('should delete category', () => {
        expect(
          vi.mocked(fetch),
        ).toHaveBeenCalledWith(
          '/api/wishlist/categories/1',
          { method: 'DELETE' },
        );
      });

      it('should refresh categories', () => {
        expect(vi.mocked(fetch)).toHaveBeenCalledWith('/api/wishlist/categories');
      });

      it('should set refreshed categories into store', async () => {
        await waitFor(
          () => expect(
            vi.mocked(categories.set),
          ).toHaveBeenCalledWith(
            [{ id: 3, name: 'category 3' }],
          ),
        );
      });
    });
  });
});
