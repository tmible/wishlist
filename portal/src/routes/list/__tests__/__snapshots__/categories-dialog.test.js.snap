// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`categories dialog > should be rendered closed 1`] = `
<body>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-expanded="false"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="closed"
      id="bits-1"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;

exports[`categories dialog > should be rendered opened 1`] = `
<body
  style="margin-right: 8px; --scrollbar-width: 1024px; overflow: hidden;"
>
  <div
    id="modal-portal"
  >
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <div
      class="modal-backdrop"
      data-dialog-overlay=""
      data-state="open"
      id="bits-5"
      style="pointer-events: auto;"
    >
      <!---->
    </div>
    
    
    
    <!---->
     
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
    <!---->
     
    <div
      aria-labelledby="bits-9"
      aria-modal="true"
      class="modal-box p-0"
      data-dialog-content=""
      data-state="open"
      id="bits-6"
      role="dialog"
      style="pointer-events: auto; outline: undefined;"
      tabindex="-1"
    >
      <div
        class="overflow-y-auto p-6"
      >
        <div
          class="flex items-start prose mb-3"
        >
          <!---->
          <div
            aria-level="2"
            class="mt-0 mr-auto"
            data-dialog-title=""
            data-state="open"
            id="bits-9"
            role="heading"
          >
            Категории
            <!---->
          </div>
          
          <!---->
           
          <!---->
          <button
            class="cursor-pointer"
            data-dialog-close=""
            data-state="open"
            id="bits-10"
            tabindex="0"
          >
            <svg
              class="lucide-icon lucide lucide-x"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <!---->
              <path
                d="M18 6 6 18"
              >
                
              </path>
              
              <!---->
              <path
                d="m6 6 12 12"
              >
                
              </path>
              
              <!---->
              <!---->
              
              <!---->
            </svg>
            <!---->
          </button>
          
          <!---->
        </div>
         
        <div
          class="flex items-center gap-4 mb-3"
        >
          <input
            class="input w-full bg-base-200"
            placeholder="Новая категория"
            type="text"
          />
           
          <button
            class="btn btn-primary btn-disabled"
            disabled=""
          >
            Добавить
          </button>
        </div>
         
        <div
          class="flex flex-col"
        >
          <div
            class="flex items-center justify-between rounded-lg hover:bg-base-200 pl-4"
          >
            category 1 
            <div>
              <button
                class="btn btn-ghost"
                data-testid="edit-button"
              >
                <svg
                  class="lucide-icon lucide lucide-pencil"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="m15 5 4 4"
                  >
                    
                  </path>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
               
              <button
                class="btn btn-ghost"
                data-testid="delete-button"
              >
                <svg
                  class="lucide-icon lucide lucide-trash-2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M3 6h18"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                  >
                    
                  </path>
                  
                  <!---->
                  <line
                    x1="10"
                    x2="10"
                    y1="11"
                    y2="17"
                  >
                    
                  </line>
                  
                  <!---->
                  <line
                    x1="14"
                    x2="14"
                    y1="11"
                    y2="17"
                  >
                    
                  </line>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
            </div>
            <!---->
          </div>
          <div
            class="flex items-center justify-between rounded-lg hover:bg-base-200 pl-4"
          >
            category 2 
            <div>
              <button
                class="btn btn-ghost"
                data-testid="edit-button"
              >
                <svg
                  class="lucide-icon lucide lucide-pencil"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="m15 5 4 4"
                  >
                    
                  </path>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
               
              <button
                class="btn btn-ghost"
                data-testid="delete-button"
              >
                <svg
                  class="lucide-icon lucide lucide-trash-2"
                  fill="none"
                  height="24"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <!---->
                  <path
                    d="M3 6h18"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
                  >
                    
                  </path>
                  
                  <!---->
                  <path
                    d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
                  >
                    
                  </path>
                  
                  <!---->
                  <line
                    x1="10"
                    x2="10"
                    y1="11"
                    y2="17"
                  >
                    
                  </line>
                  
                  <!---->
                  <line
                    x1="14"
                    x2="14"
                    y1="11"
                    y2="17"
                  >
                    
                  </line>
                  
                  <!---->
                  <!---->
                  
                  <!---->
                </svg>
                <!---->
              </button>
            </div>
            <!---->
          </div>
          
        </div>
      </div>
      <!---->
    </div>
    
    
    
    
    
    
    
    <!---->
    
    
    
  </div>
  <div>
    <!---->
    <!---->
    <!---->
    <!---->
    <button
      aria-controls="bits-6"
      aria-expanded="true"
      aria-haspopup="dialog"
      class="hidden"
      data-dialog-trigger=""
      data-state="open"
      id="bits-4"
    >
      <!---->
    </button>
    
    <!---->
     
    <!---->
    
    <!---->
    
    
    
  </div>
</body>
`;
